pilot/pkg/serviceregistry/kube/controller/ambientindex_test.go

```diff
@@ -240,27 +243,54 @@ func TestAmbientIndex_WaypointAddressAddedToWorkloads(t *testing.T) {
 			constants.ManagedGatewayLabel: constants.ManagedGatewayMeshControllerLabel,
 			constants.GatewayNameLabel:    "namespace-wide",
 		}, nil, true, corev1.PodRunning)
-	s.assertAddresses(t, "", "pod1", "pod2", "pod3", "waypoint-ns-pod")
+	// Add a waypoint proxy pod for sa2
+	s.addPods(t, "127.0.0.201", "waypoint-sa2-pod", "service-account",
+		map[string]string{
+			constants.ManagedGatewayLabel: constants.ManagedGatewayMeshControllerLabel,
+			constants.GatewayNameLabel:    "service-account",
+		}, nil, true, corev1.PodRunning)
+	s.assertAddresses(t, "", "pod1", "pod2", "pod3", "pod4", "waypoint-ns-pod", "waypoint-sa2-pod")
 	s.assertEvent(t, s.podXdsName("waypoint-ns-pod"))
 
+	s.addWaypoint(t, "10.0.0.2", "waypoint-ns", "")
 	// create the waypoint service
 	s.addService(t, "waypoint-ns",
 		map[string]string{constants.ManagedGatewayLabel: constants.ManagedGatewayMeshControllerLabel},
 		map[string]string{},
 		[]int32{80}, map[string]string{constants.GatewayNameLabel: "namespace-wide"}, "10.0.0.2")
-	s.assertAddresses(t, "", "pod1", "pod2", "pod3", "waypoint-ns", "waypoint-ns-pod")
+
+	s.addWaypoint(t, "10.0.0.3", "waypoint-sa2", "sa2")
+	// create the waypoint service
+	s.addService(t, "waypoint-sa2",
+		map[string]string{constants.ManagedGatewayLabel: constants.ManagedGatewayMeshControllerLabel},
+		map[string]string{},
+		[]int32{80}, map[string]string{constants.GatewayNameLabel: "service-account"}, "10.0.0.3")
+
+	s.assertAddresses(t, "", "pod1", "pod2", "pod3", "pod4", "waypoint-ns", "waypoint-ns-pod", "waypoint-sa2-pod", "waypoint-sa2")
 	// All these workloads updated, so push them
 	s.assertEvent(t, s.podXdsName("pod1"),
 		s.podXdsName("pod2"),
 		s.podXdsName("pod3"),
+		s.podXdsName("pod4"),
+	)
+	s.assertEvent(t,
 		s.podXdsName("waypoint-ns-pod"),
 		s.svcXdsName("waypoint-ns"),
 	)
+	s.assertEvent(t,
+		s.podXdsName("waypoint-sa2-pod"),
+		s.svcXdsName("waypoint-sa2"),
+	)
+
```

Comment:
It seems unlikely the ordering of these is guaranteed?

Reply:
I had a similar issue in https://github.com/istio/istio/pull/45426/files and added some other checkers

Reply:
Interestingly, I did not have trouble with this when using the `configController` event handler but after switching to `gatewayResourceClient` unit tests are failing with what looks to be ordering issues quite regularly.